HW2/P4/driver.py

```diff
@@ -16,20 +16,43 @@
 import threading
 
 def py_median_3x3(image, iterations=10, num_threads=1):
-    ''' repeatedly filter with a 3x3 median '''
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
 
-    for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
-        tmpA, tmpB = tmpB, tmpA
+    # thread i controls every ith line of image
+    threads = [None] * num_threads
+
+    # create event where each corresponds to whether each thread and then iteration is done
+    events = [[threading.Event() for i in xrange(iterations)] for j in xrange(num_threads)]
+
+    for t_id in xrange(num_threads):
+        threads[t_id] = threading.Thread(target=filter_image, args=(num_threads, t_id, events, iterations, tmpA, tmpB))
+        threads[t_id].start()
+
+    # terminate threads
+    for thread in threads:
+        thread.join()
 
     return tmpA
 
+def filter_image(num_threads, t_id, events, iterations, tmpA, tmpB):
+    for i in xrange(iterations):
+        # makes the i-1 iteration of t_id-1 and t_id+1 wait before doing iteration i
+        if num_threads > 1 and i > 0:
+            events[(t_id-1) % num_threads][i-1].wait()
+            events[(t_id+1) % num_threads][i-1].wait()
+
+        filtering.median_3x3(tmpA, tmpB, t_id, num_threads)
+
+        # set flag
+        events[t_id][i].set()
+
```

Comment:
I used the same approach to complete the extra credit part by creating an 2-D array of events!